crates/walrus-service/src/node/blob_event_processor.rs

```diff
@@ -0,0 +1,311 @@
+            // subsequent certify or delete events may update the `blob_info`; so we cannot remove
+            // it even if it is no longer valid in the *current* epoch
+            if !blob_info.is_registered(event.epoch) {
+                tracing::debug!("deleting data for deleted blob");
+                // TODO (WAL-201): Actually delete blob data.
+            }
+        } else {
+            tracing::warn!(
+                walrus.blob_id = %blob_id,
+                "handling `BlobDeleted` event for an untracked blob"
+            );
+        }
+
+        event_handle.mark_as_complete();
+
+        Ok(())
+    }
+
+    /// Processes a blob invalid event.
+    #[tracing::instrument(skip_all)]
+    async fn process_blob_invalid_event(
+        &self,
+        event_handle: EventHandle,
+        event: InvalidBlobId,
+    ) -> anyhow::Result<()> {
+        self.node
+            .blob_retirement_notifier
+            .notify_blob_retirement(&event.blob_id);
+        self.blob_sync_handler
+            .cancel_sync_and_mark_event_complete(&event.blob_id)
+            .await?;
+        self.node.storage.delete_blob_data(&event.blob_id).await?;
+
+        event_handle.mark_as_complete();
+        Ok(())
+    }
+}
+
+/// Blob event processor that processes blob events. It can be configured to process events
+/// sequentially or in parallel using background workers.
+#[derive(Debug, Clone)]
+pub struct BlobEventProcessor {
+    node: Arc<StorageNodeInner>,
+
+    // Background processors that process events in parallel.
+    background_processor_senders: Vec<UnboundedSender<(EventHandle, BlobEvent)>>,
+    _background_processors: Vec<Arc<JoinHandle<()>>>,
+
+    // When there are no background workers, we use a sequential processor to process events using
+    // this processor. This is to keep the same behavior as before BackgroundEventProcessor.
+    sequential_processor: Option<Arc<BackgroundEventProcessor>>,
+}
+
+impl BlobEventProcessor {
+    pub fn new(
+        node: Arc<StorageNodeInner>,
+        blob_sync_handler: Arc<BlobSyncHandler>,
+        num_workers: usize,
+    ) -> Self {
+        let mut senders = Vec::with_capacity(num_workers);
+        let mut workers = Vec::with_capacity(num_workers);
+        for worker_index in 0..num_workers {
+            let (tx, rx) = mpsc::unbounded_channel();
+            senders.push(tx);
+            let mut background_processor =
+                BackgroundEventProcessor::new(node.clone(), blob_sync_handler.clone(), rx);
+            workers.push(Arc::new(tokio::spawn(async move {
+                background_processor.run(worker_index).await;
+            })));
+        }
+
+        let sequential_processor = if num_workers == 0 {
+            // Create a sequential processor to process events sequentially if no background workers
+            // are configured.
+            let (_tx, rx) = mpsc::unbounded_channel();
```

Comment:
Nit: It feels a bit unclean to have this receiver that isn't actually used. I think it's fine though, as we probably will remove the sequential processing altogether soon, right?